File: apex/cloud/kv.py
"""Minimal Upstash/Vercel-KV REST client (no extra dependency — uses ``requests``).

Vercel KV is Upstash Redis under the hood and exposes a REST API. We read the same
env vars Vercel injects (``KV_REST_API_URL`` / ``KV_REST_API_TOKEN``) or the native
Upstash names. Values are stored as JSON strings, compatible with the dashboard's
``@vercel/kv`` client (which JSON-stringifies on set and parses on get).

Every call is best-effort and never raises: on any error it logs and returns a safe
default, so a KV blip can never crash the trading loop.
"""

from __future__ import annotations

import json
import os
from typing import Any

import requests
from loguru import logger

CONFIG_KEY = "apex:config"            # onboarding config written by the dashboard
STATE_KEY = "apex:state"              # live snapshot pushed by the laptop
STATUS_KEY = "apex:onboarding_status" # algo-confirmed onboarding status
BACKTEST_REQ_KEY = "apex:backtest_request"   # dashboard → laptop: run this backtest
BACKTEST_RES_KEY = "apex:backtest_result"    # laptop → dashboard: backtest result

_TIMEOUT = 8


def _creds() -> tuple[str | None, str | None]:
    url = os.getenv("KV_REST_API_URL") or os.getenv("UPSTASH_REDIS_REST_URL")
    token = os.getenv("KV_REST_API_TOKEN") or os.getenv("UPSTASH_REDIS_REST_TOKEN")
    if url and token:
        return url.rstrip("/"), token
    return None, None


def kv_enabled() -> bool:
    """True when KV credentials are present (cloud-relay mode is active)."""
    url, token = _creds()
    return bool(url and token)


def _headers(token: str) -> dict[str, str]:
    return {"Authorization": f"Bearer {token}"}


def kv_get(key: str) -> Any | None:
    """Return the JSON-decoded value at ``key``, or ``None`` if absent/unreachable."""
    url, token = _creds()
    if not url or not token:
        return None
    try:
        resp = requests.get(f"{url}/get/{key}", headers=_headers(token), timeout=_TIMEOUT)
        resp.raise_for_status()
        result = resp.json().get("result")
        if result in (None, ""):
            return None
        return json.loads(result) if isinstance(result, str) else result
    except Exception as exc:
        logger.warning("KV get '{}' failed: {}", key, exc)
        return None


def kv_set(key: str, value: Any) -> bool:
    """Store ``value`` (JSON-encoded) at ``key``. Returns success."""
    url, token = _creds()
    if not url or not token:
        return False
    try:
        body = json.dumps(value)
        resp = requests.post(f"{url}/set/{key}", headers=_headers(token), data=body, timeout=_TIMEOUT)
        resp.raise_for_status()
        return True
    except Exception as exc:
        logger.warning("KV set '{}' failed: {}", key, exc)
        return False


def kv_delete(key: str) -> bool:
    url, token = _creds()
    if not url or not token:
        return False
    try:
        resp = requests.post(f"{url}/del/{key}", headers=_headers(token), timeout=_TIMEOUT)
        resp.raise_for_status()
        return True
    except Exception as exc:
        logger.warning("KV del '{}' failed: {}", key, exc)
        return False

File: dashboard/lib/kv.ts
// Minimal Upstash/Vercel-KV REST client (no SDK — uses fetch). Accepts either the
// Vercel-injected KV_REST_API_* names or the native UPSTASH_REDIS_REST_* names, so it
// works whichever Redis integration you attach. Values are JSON strings, matching
// apex/cloud/kv.py exactly (laptop and dashboard share the same keys).

export const CONFIG_KEY = "apex:config";
export const STATE_KEY = "apex:state";
export const STATUS_KEY = "apex:onboarding_status";
export const BACKTEST_REQ_KEY = "apex:backtest_request";
export const BACKTEST_RES_KEY = "apex:backtest_result";

function creds(): { url: string; token: string } | null {
  const url = process.env.KV_REST_API_URL || process.env.UPSTASH_REDIS_REST_URL;
  const token = process.env.KV_REST_API_TOKEN || process.env.UPSTASH_REDIS_REST_TOKEN;
  if (!url || !token) return null;
  return { url: url.replace(/\/$/, ""), token };
}

/** Cloud-relay mode is active whenever Redis REST credentials are present. */
export function kvEnabled(): boolean {
  return creds() !== null;
}

export async function kvGet<T>(key: string): Promise<T | null> {
  const c = creds();
  if (!c) return null;
  try {
    const res = await fetch(`${c.url}/get/${key}`, {
      headers: { Authorization: `Bearer ${c.token}` },
      cache: "no-store",
    });
    if (!res.ok) return null;
    const body = (await res.json()) as { result: string | null };
    if (body.result == null || body.result === "") return null;
    return JSON.parse(body.result) as T;
  } catch {
    return null;
  }
}

export async function kvSet(key: string, value: unknown): Promise<boolean> {
  const c = creds();
  if (!c) return false;
  try {
    const res = await fetch(`${c.url}/set/${key}`, {
      method: "POST",
      headers: { Authorization: `Bearer ${c.token}` },
      body: JSON.stringify(value),
      cache: "no-store",
    });
    return res.ok;
  } catch {
    return false;
  }
}

export async function kvDel(key: string): Promise<boolean> {
  const c = creds();
  if (!c) return false;
  try {
    const res = await fetch(`${c.url}/del/${key}`, {
      method: "POST",
      headers: { Authorization: `Bearer ${c.token}` },
      cache: "no-store",
    });
    return res.ok;
  } catch {
    return false;
  }
}
